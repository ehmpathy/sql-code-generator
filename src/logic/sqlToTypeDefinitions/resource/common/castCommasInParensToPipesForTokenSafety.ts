export const castCommasInParensToPipesForTokenSafety = ({ sql }: { sql: string }) => {
  // track state of whether we are in parenthesizes or not
  let areInsideParensDepth = 0; // init as 0, since not inside parens
  const areInsideParens = () => areInsideParensDepth > 0; // whenever areInsideParensDepth > 0, we are inside parens

  // init object to build up the "casted" string
  let castedString = '';

  // loop through each char and update state / replace char as needed
  [...sql].forEach((char) => {
    // 1. manage "areInsideParens" state
    if (char === '(') {
      areInsideParensDepth += 1; // one parens deeper!
    }
    if (char === ')') {
      if (!areInsideParens) throw new Error('exiting parenthesizes without having opened them; not supported'); // fail fast for unsupported case
      areInsideParensDepth -= 1; // mark that we've dropped off one parens now
    }

    // 2. cast the char, considering whether we're in parens or not
    const castedChar = areInsideParens() && char === ',' ? '|' : char; // cast "," if we are inside parens

    // 3. add casted char to string
    castedString += castedChar;
  });

  // return the casted string
  return castedString;
};
