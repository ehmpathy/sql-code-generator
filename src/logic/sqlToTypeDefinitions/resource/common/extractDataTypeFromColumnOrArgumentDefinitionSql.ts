import { DataType } from '../../../../model';

// https://dev.mysql.com/doc/refman/8.0/en/string-types.html
const mysqlStringTypes = [
  'CHAR',
  'VARCHAR',
  'TINYTEXT',
  'TEXT',
  'MEDIUMTEXT',
  'LONGTEXT',
  'ENUM', // note: we treat enums as strings; TODO: https://github.com/uladkasach/sql-code-generator/issues/1
];

// https://www.postgresql.org/docs/9.5/datatype-character.html
const postgresStringTypes = [
  'CHARACTER',
  'CHAR',
  'CHARACTER VARYING',
  'VARCHAR',
  'TEXT',
  'UUID', // this too: https://www.postgresql.org/docs/9.5/datatype-uuid.html
];

// combine the types so we can search over the union
const dbStringTypes = [...mysqlStringTypes, ...postgresStringTypes];

// https://dev.mysql.com/doc/refman/8.0/en/numeric-types.html
const mysqlNumberTypes = [
  'INTEGER',
  'INT',
  'SMALLINT',
  'TINYINT',
  'MEDIUMINT',
  'BIGINT',
  'DECIMAL',
  'NUMERIC',
  'FLOAT',
  'DOUBLE',
];

// https://www.postgresql.org/docs/9.5/datatype-numeric.html
const postgresNumberTypes = [
  'SMALLINT',
  'INT2',
  'INTEGER',
  'INT',
  'INT4',
  'BIGINT',
  'INT8',
  'DECIMAL',
  'NUMERIC',
  'REAL',
  'DOUBLE PRECISION',
  'FLOAT8',
  'SMALLSERIAL',
  'SERIAL2',
  'SERIAL',
  'SERIAL4',
  'BIGSERIAL',
  'SERIAL8',
];

// combine the types so we can search over the union
const dbNumberTypes = [...mysqlNumberTypes, ...postgresNumberTypes];

// https://dev.mysql.com/doc/refman/8.0/en/date-and-time-types.html
const mysqlDateTypes = ['DATE', 'TIME', 'DATETIME', 'TIMESTAMP', 'YEAR'];

// https://www.postgresql.org/docs/9.5/datatype-datetime.html
const postgresDateTypes = [
  'TIMESTAMP',
  'TIMESTAMPTZ',
  'TIMESTAMP WITH TIMEZONE',
  'DATE',
  'TIME',
  'TIMETZ',
  'TIME WITH TIMEZONE',
];

// combine the types so we can search over the union
const dbDateTypes = [...mysqlDateTypes, ...postgresDateTypes];

// https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
const mysqlBinaryTypes = ['BINARY', 'VARBINARY'];

// https://www.postgresql.org/docs/9.5/datatype-binary.html
const postgresBinaryTypes = ['BYTEA'];

// combine the types so we can search over the union
const dbBinaryTypes = [...mysqlBinaryTypes, ...postgresBinaryTypes];

export const extractDataTypeFromColumnOrArgumentDefinitionSql = ({ sql }: { sql: string }) => {
  // 1. cast to upper
  const sqlUpper = sql.toUpperCase();

  // 2. one by one try to see which datatype is being defined in this definition
  if (dbStringTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.STRING;
  if (dbNumberTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.NUMBER;
  if (dbDateTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.DATE;
  if (dbBinaryTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.BUFFER;
  throw new Error(`could not extract data type from '${sql}'`);
};
