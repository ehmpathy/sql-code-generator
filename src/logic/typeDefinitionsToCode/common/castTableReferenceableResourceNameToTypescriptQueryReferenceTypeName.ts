import { ResourceType } from '../../../model';
import { castResourceNameToTypescriptTypeName } from './castResourceNameToTypescriptTypeName';

export const castTableReferenceableResourceNameToTypescriptQueryReferenceTypeName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  // 0. throw error if not view or table, only view and table are possible table references
  if (![ResourceType.TABLE, ResourceType.VIEW].includes(resourceType)) {
    throw new Error('not a table referenceable query');
  }

  // 1. get the regular typescript type name
  const typescriptTypeName = castResourceNameToTypescriptTypeName({ name, resourceType });

  // 2. return the referenceable name
  if (resourceType === ResourceType.TABLE) return typescriptTypeName.replace(/^SqlTable/, 'SqlTableOrView');
  if (resourceType === ResourceType.VIEW) return typescriptTypeName.replace(/^SqlView/, 'SqlTableOrView');
  throw new Error('unexpected'); // this should never occur, but fail fast
};
