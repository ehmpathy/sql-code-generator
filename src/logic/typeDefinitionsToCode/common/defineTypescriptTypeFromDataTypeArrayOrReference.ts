import {
  DataType,
  TypeDefinitionReference,
  TypeDefinition,
  TypeDefinitionOfQueryTableReference,
} from '../../../domain';
import { defineTypescriptTypeFromDataTypeArray } from './defineTypescriptTypeFromDataTypeArray';
import { defineTypescriptTypeFromReference } from './defineTypescriptTypeFromReference/defineTypescriptTypeFromReference';

export const defineTypescriptTypeFromDataTypeArrayOrReference = ({
  type,
  plural,
  queryTableReferences,
  typeDefinitions,
}: {
  type: DataType[] | TypeDefinitionReference;
  plural: boolean;
  typeDefinitions: TypeDefinition[];
  queryTableReferences: TypeDefinitionOfQueryTableReference[];
}) => {
  // if its a type reference, then use that handler
  if (type instanceof TypeDefinitionReference) {
    const defined = defineTypescriptTypeFromReference({
      reference: type,
      queryTableReferences,
      typeDefinitions,
    });
    if (plural) return `${defined}[]`;
    return defined;
  }

  // else, it must be data array. use that handler
  return defineTypescriptTypeFromDataTypeArray({ type });
};
