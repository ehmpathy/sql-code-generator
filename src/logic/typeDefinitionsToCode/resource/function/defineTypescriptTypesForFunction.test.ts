import { extractSqlFromFile } from '../../../common/extractSqlFromFile';
import { extractTypeDefinitionFromFunctionSql } from '../../../sqlToTypeDefinitions/resource/function/extractTypeDefinitionFromFunctionSql';
import { defineTypescriptTypesForFunction } from './defineTypescriptTypesForFunction';

describe('defineTypescriptTypesForFunction', () => {
  it('should generate accurate looking types for this function definition', async () => {
    const definition = extractTypeDefinitionFromFunctionSql({
      name: 'upsert_image',
      sql: await extractSqlFromFile({
        filePath: `${__dirname}/../../../__test_assets__/functions/upsert_image.mysql.sql`,
      }),
    });
    const code = defineTypescriptTypesForFunction({ definition });
    expect(code).toMatchSnapshot();
  });
  it('should generate accurate looking types for this other function definition', async () => {
    const definition = extractTypeDefinitionFromFunctionSql({
      name: 'upsert_suggestion',
      sql: await extractSqlFromFile({
        filePath: `${__dirname}/../../../__test_assets__/functions/upsert_suggestion.mysql.sql`,
      }),
    });
    const code = defineTypescriptTypesForFunction({ definition });
    expect(code).toMatchSnapshot();
  });
  it('should generate accurate looking types for this other other function definition', async () => {
    const definition = extractTypeDefinitionFromFunctionSql({
      name: 'hash_string',
      sql: await extractSqlFromFile({
        filePath: `${__dirname}/../../../__test_assets__/functions/hash_string.mysql.sql`,
      }),
    });
    const code = defineTypescriptTypesForFunction({ definition });
    expect(code).toMatchSnapshot();
  });
  it('should generate accurate looking types for this other other other function definition', async () => {
    const definition = extractTypeDefinitionFromFunctionSql({
      name: 'upsert_photo',
      sql: await extractSqlFromFile({
        filePath: `${__dirname}/../../../__test_assets__/functions/upsert_photo.postgres.sql`,
      }),
    });
    const code = defineTypescriptTypesForFunction({ definition });
    expect(code).toMatchSnapshot();
  });
  it('should generate accurate looking types for this other function definition, with array inputs', async () => {
    const definition = extractTypeDefinitionFromFunctionSql({
      name: 'upsert_job',
      sql: await extractSqlFromFile({
        filePath: `${__dirname}/../../../__test_assets__/functions/upsert_job.postgres.sql`,
      }),
    });
    const code = defineTypescriptTypesForFunction({ definition });
    expect(code).toMatchSnapshot();
  });
});
