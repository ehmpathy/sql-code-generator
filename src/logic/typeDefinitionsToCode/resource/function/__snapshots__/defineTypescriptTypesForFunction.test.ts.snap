// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineTypescriptTypesForFunction should generate accurate looking types for this function definition 1`] = `
"export interface SqlFunctionUpsertImageInput {
  0: string | null;
  1: string | null;
  2: string | null;
  3: string | null;
}
export interface SqlFunctionUpsertImageInputByName {
  in_url: SqlFunctionUpsertImageInput['0'];
  in_caption: SqlFunctionUpsertImageInput['1'];
  in_credit: SqlFunctionUpsertImageInput['2'];
  in_alt_text: SqlFunctionUpsertImageInput['3'];
}
export type SqlFunctionUpsertImageOutput = number;"
`;

exports[`defineTypescriptTypesForFunction should generate accurate looking types for this other function definition 1`] = `
"export interface SqlFunctionUpsertSuggestionInput {
  0: string | null;
  1: string | null;
  2: number | null;
  3: string | null;
  4: string | null;
  5: number | null;
}
export interface SqlFunctionUpsertSuggestionInputByName {
  in_suggestion_source: SqlFunctionUpsertSuggestionInput['0'];
  in_external_id: SqlFunctionUpsertSuggestionInput['1'];
  in_suggested_idea_id: SqlFunctionUpsertSuggestionInput['2'];
  in_status: SqlFunctionUpsertSuggestionInput['3'];
  in_result: SqlFunctionUpsertSuggestionInput['4'];
  in_resultant_curated_idea_id: SqlFunctionUpsertSuggestionInput['5'];
}
export type SqlFunctionUpsertSuggestionOutput = number;"
`;

exports[`defineTypescriptTypesForFunction should generate accurate looking types for this other function definition, with array inputs 1`] = `
"export interface SqlFunctionUpsertJobInput {
  0: string | null;
  1: string | null;
  2: string | null;
  3: string | null;
  4: number | null;
  5: number[] | null;
}
export interface SqlFunctionUpsertJobInputByName {
  in_uuid: SqlFunctionUpsertJobInput['0'];
  in_title: SqlFunctionUpsertJobInput['1'];
  in_status: SqlFunctionUpsertJobInput['2'];
  in_description: SqlFunctionUpsertJobInput['3'];
  in_location_id: SqlFunctionUpsertJobInput['4'];
  in_photo_ids: SqlFunctionUpsertJobInput['5'];
}
export type SqlFunctionUpsertJobOutput = number;"
`;

exports[`defineTypescriptTypesForFunction should generate accurate looking types for this other other function definition 1`] = `
"export interface SqlFunctionHashStringInput {
  0: string | null;
  1: string | null;
  2: number | null;
}
export interface SqlFunctionHashStringInputByName {
  in_array: SqlFunctionHashStringInput['0'];
  in_delimiter: SqlFunctionHashStringInput['1'];
  in_index: SqlFunctionHashStringInput['2'];
}
export type SqlFunctionHashStringOutput = string;"
`;

exports[`defineTypescriptTypesForFunction should generate accurate looking types for this other other other function definition 1`] = `
"export interface SqlFunctionUpsertPhotoInput {
  0: string | null;
  1: string | null;
}
export interface SqlFunctionUpsertPhotoInputByName {
  in_url: SqlFunctionUpsertPhotoInput['0'];
  in_description: SqlFunctionUpsertPhotoInput['1'];
}
export type SqlFunctionUpsertPhotoOutput = number;"
`;
